File: solutions/scala/secret-handshake/1/src/main/scala/SecretHandshake.scala
object SecretHandshake {
  private val Reverse = 16

  private val actions = List(
    1 -> "wink",
    2 -> "double blink",
    4 -> "close your eyes",
    8 -> "jump"
  )

  def commands(number: Int): List[String] =
    val selectedActions =
      actions
        .collect { case (bit, action) if hasBit(number, bit) => action }

    if hasBit(number, Reverse) then selectedActions.reverse
    else selectedActions

  private def hasBit(number: Int, bit: Int): Boolean =
    (number & bit) != 0
}

File: solutions/fsharp/secret-handshake/1/SecretHandshake.fs
module SecretHandshake

let private actions =
    [
        1, "wink"
        2, "double blink"
        4, "close your eyes"
        8, "jump"
    ]

let commands number =
    let selected =
        actions
        |> List.choose (fun (mask, action) ->
            if number &&& mask <> 0 then Some action else None)

    if number &&& 16 <> 0 then List.rev selected else selected
